apps/src/common.rs

```diff
@@ -1640,3 +1649,1002 @@ impl HttpConn for Http3Conn {
+                    break;
+                },
+
+                Err(e) => {
+                    error!("{} HTTP/3 error {:?}", conn.trace_id(), e);
+
+                    return Err(e);
+                },
+            }
+        }
+
+        // Process datagram-related events.
+        while let Ok(len) = conn.dgram_recv(buf) {
+            let mut b = octets::Octets::with_slice(buf);
+            if let Ok(flow_id) = b.get_varint() {
+                info!(
+                    "Received DATAGRAM flow_id={} len={} data={:?}",
+                    flow_id,
+                    len,
+                    buf[b.off()..len].to_vec()
+                );
+            }
+        }
+
+        if let Some(ds) = self.dgram_sender.as_mut() {
+            let mut dgrams_done = 0;
+
+            for _ in ds.dgrams_sent..ds.dgram_count {
+                match send_h3_dgram(conn,conn_paths, ds.flow_id, ds.dgram_content.as_bytes())
+                {
+                    Ok(v) => v,
+
+                    Err(e) => {
+                        error!("failed to send dgram {:?}", e);
+                        break;
+                    },
+                }
+
+                dgrams_done += 1;
+            }
+
+            ds.dgrams_sent += dgrams_done;
+        }
+
+        Ok(())
+    }
+
+    fn handle_writable(
+        &mut self, conn: &mut quiche::Connection,
+        partial_responses: &mut HashMap<u64, PartialResponse>, stream_id: u64,
+    ) {
+        debug!("{} stream {} is writable", conn.trace_id(), stream_id);
+
+        if !partial_responses.contains_key(&stream_id) {
+            return;
+        }
+
+        let resp = partial_responses.get_mut(&stream_id).unwrap();
+
+        if let (Some(headers), Some(priority)) = (&resp.headers, &resp.priority) {
+            match self.h3_conn.send_response_with_priority(
+                conn, stream_id, headers, priority, false,
+            ) {
+                Ok(_) => (),
+
+                Err(quiche::h3::Error::StreamBlocked) => {
+                    return;
+                },
+
+                Err(e) => {
+                    error!("{} stream send failed {:?}", conn.trace_id(), e);
+                    return;
+                },
+            }
+        }
+
+        resp.headers = None;
+        resp.priority = None;
+
+        let body = &resp.body[resp.written..];
+
+        let written = match self.h3_conn.send_body(conn, stream_id, body, true) {
+            Ok(v) => v,
+
+            Err(quiche::h3::Error::Done) => 0,
+
+            Err(e) => {
+                partial_responses.remove(&stream_id);
+
+                error!("{} stream send failed {:?}", conn.trace_id(), e);
+                return;
+            },
+        };
+
+        resp.written += written;
+
+        if resp.written == resp.body.len() {
+            partial_responses.remove(&stream_id);
+        }
+    }
+}
+
+
+
+// The code below has been added by: Vany Ingenzi
+pub struct Scheduler {
```

Comment:
Looking back, why do we need a scheduler? Shouldn't each path try to send as much as they can when app data is available?